### PR TITLE
Environment fixes

### PR DESCRIPTION
**Pull request by bobhy**
Proposed fix for #3739 and #3703
 
Allows you to specify default of `DefaultNotGiven when registering an environment variable, which means raise KeyError rather than return the default if that variable not actually defined in environment.  Used that for "ANSICON", which fixes the immediate problem, and for "TERM" and "VIRTUAL_ENV", reverting some recent changes to match).  The point is that these environment variables (and maybe OLDPWD?) are conventionally expected to be defined outside of Xonsh, so whether they exist is significant.  

No change in this patch to evaluation of non-registered environment variables when they aren't defined in environment, that's a topic for another day.

### CODE REVIEW

File: news/environment_fixes.rst [hunk 1]
```diff
@@ -0,0 +1,30 @@
+**Added:**
+
+* <news item>
+
+**Changed:**
+
+* xonsh/environ.py: new rule: if default for "registered" environment variable is set to ``DefaultNotGiven``_
```
Comment (scopatz) [inline news/environment_fixes.rst:7]: I think that the trailing underscore here should be removed in the in-line code

File: tests/test_environ.py [hunk 2]
```diff
@@ -357,7 +355,7 @@ def test_register_custom_var_generic():
     assert env["MY_SPECIAL_VAR"] == 32
 
     env["MY_SPECIAL_VAR"] = True
-    assert env["MY_SPECIAL_VAR"] == True
+    assert env["MY_SPECIAL_VAR"] == True  # NOQA E712
```
Comment (scopatz) [inline tests/test_environ.py:358]: probably should either be `assert env["MY_SPECIAL_VAR"] is True` or `assert env["MY_SPECIAL_VAR"]`

File: tests/tools.py [hunk 3]
```diff
@@ -109,7 +109,7 @@ class DummyEnv(MutableMapping):
     DEFAULTS = {
         "XONSH_DEBUG": 1,
         "XONSH_COLOR_STYLE": "default",
-        "VIRTUAL_ENV": "",
+        # "VIRTUAL_ENV": "",
```
Comment (scopatz) [inline tests/tools.py:112]: Please remove dead code

**Comment by scopatz** [comment]
Thanks @bobhy - just a couple of small updates

File: news/environment_fixes.rst (hunk 1 above)
Comment (bobhy) [reply news/environment_fixes.rst:7]: ok

File: tests/test_environ.py (hunk 2 above)
Comment (bobhy) [reply tests/test_environ.py:358]: ok

File: tests/tools.py (hunk 3 above)
Comment (bobhy) [reply tests/tools.py:112]: ok

Comment (bobhy) [reply tests/tools.py:112]: ok

**Comment by scopatz** [comment]
It looks like there are conflicts here now, since I merged the other branch

**Comment by bobhy** [comment]
This is ready for review, and I'd like to get it resolved, one way or the other.

**Comment by scopatz** [comment]
Thanks!

**Pull request merged**